### PR TITLE
Setup Azure Storage with Terraform

### PR DESCRIPTION
**Pull request by KarolJastrzebski**
https://tools.hmcts.net/jira/browse/RPE-439

This sets up single container for all the scans.

`Jenkinsfile_parametrized` used to play around in sandbox environment.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: Jenkinsfile_parameterized [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#!groovy
+@Library(value="Infrastructure", changelog=false) _
+
+properties([
+  parameters([
+    string(name: 'PRODUCT_NAME', defaultValue: 'bulk-scanning', description: ''),
+    string(name: 'APP', defaultValue: 'bulk-scanning', description: ''),
+    string(name: 'TYPE', defaultValue: 'java', description: ''),
+    string(name: 'ENVIRONMENT', defaultValue: 'sandbox', description: 'Environment where code should be build and deployed'),
```
Comment (banderous) [inline Jenkinsfile_parameterized]: built

File: infrastructure/main.tf [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  source       = "git@github.com:hmcts/moj-module-webapp?ref=master"
+  product      = "${var.product}-${local.app}"
+  location     = "${var.location}"
+  env          = "${var.env}"
+  ilbIp        = "${var.ilbIp}"
+  subscription = "${var.subscription}"
+  is_frontend  = false
+  capacity     = "${var.capacity}"
+
+  asp_name = "bulk-scanning"
```
Comment (banderous) [inline infrastructure/main.tf]: asp_name?

File: infrastructure/main.tf [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+resource "azurerm_storage_account" "provider" {
+  name                     = "bulkscanning"
+  resource_group_name      = "${azurerm_resource_group.rg.name}"
+  location                 = "${var.location}"
+  account_tier             = "Standard"
+  account_replication_type = "LRS"
+}
+
+resource "azurerm_storage_container" "scans" {
```
Comment (banderous) [inline infrastructure/main.tf]: apparently we might receive more than just scans, such as information about cash that has been sent in. Maybe call it something like 'inbound'?

File: infrastructure/prod.tfvars [hunk 4]
```diff
@@ -0,0 +1 @@
+capacity = "2"
```
Comment (banderous) [inline infrastructure/prod.tfvars]: Is this the number of app servers we'll have? Perhaps we could just have 1, it'd make writing the ingestion cron easier.

File: infrastructure/main.tf (hunk 2 above)
Comment (KarolJastrzebski) [reply infrastructure/main.tf]: Yes, copied over from rhubarb service. The variable then is used as part of the `asp_name` in deployment template.

**Comment by banderous** [comment]
Let's try and write a functional test that checks the container is provisioned.

File: infrastructure/prod.tfvars (hunk 4 above)
Comment (KarolJastrzebski) [reply infrastructure/prod.tfvars]: I wouldn't worry for now as Azure Storage has locking mechanism built in. We can change it later.

File: src/main/java/uk/gov/hmcts/reform/bulkscanning/info/BlobContainerInfoContributor.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            "container_name", container.getName(),
+            "container_exists", doesContainerExist()
+        ));
+    }
+
+    private boolean doesContainerExist() {
+        try {
+            return container.exists();
+        } catch (StorageException e) {
+            return false;
```
Comment (banderous) [inline src/main/java/uk/gov/hmcts/reform/bulkscanning/info/BlobContainerInfoContributor.java]: we should log this exception. Possibly we could put the exception in the info...

File: src/smokeTest/java/uk/gov/hmcts/reform/bulkscanning/WhenApplicationDeployedTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Value("${test-url}")
+    private String testUrl;
+
+    @Test
+    public void storage_container_exists() {
+        RestAssured
+            .given()
+            .relaxedHTTPSValidation()
+            .baseUri(testUrl)
```
Comment (banderous) [inline src/smokeTest/java/uk/gov/hmcts/reform/bulkscanning/WhenApplicationDeployedTest.java]: I would hard code testUrl or make it a constant somewhere

File: Jenkinsfile_parameterized [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#!groovy
+@Library(value="Infrastructure", changelog=false) _
+
+properties([
+  parameters([
+    string(name: 'PRODUCT_NAME', defaultValue: 'bulk-scanning', description: ''),
```
Comment (nitinprabhu) [inline Jenkinsfile_parameterized]: I think product should be `rpe` ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanning/config/StorageConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        return new StorageCredentialsAccountAndKey(accountName, key);
+    }
+
+    @Bean
+    public CloudStorageAccount getCloudStorageAccount(StorageCredentials storageCredentials) throws URISyntaxException {
+        return new CloudStorageAccount(storageCredentials, true);
+    }
+
+    @Bean
+    public CloudBlobContainer getCloudBlobContainer(
```
[the remark this answers is not in the document]
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/bulkscanning/config/StorageConfiguration.java]: You're right. Such object could also handle discrete tasks, like lock/release, or purging processed files.

I'll be adding this later on as we get better perspective.

File: Jenkinsfile_parameterized [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    string(name: 'TYPE', defaultValue: 'java', description: ''),
+    string(name: 'ENVIRONMENT', defaultValue: 'sandbox', description: 'Environment where code should be built and deployed'),
+    choice(name: 'SUBSCRIPTION', choices: 'sandbox\nnonprod\nprod', description: 'Azure subscriptions available to build in')
+  ]),
+  [$class: 'GithubProjectProperty', projectUrlStr: 'github.com/hmcts/bulk-scanning'],
+  pipelineTriggers([[$class: 'GitHubPushTrigger']])
+])
+
+withParameterizedPipeline(params.TYPE, params.PRODUCT_NAME, params.APP, params.ENVIRONMENT, params.SUBSCRIPTION) {
+
```
Comment (nitinprabhu) [inline Jenkinsfile_parameterized]: You could add  ` enableSlackNotifications(channel)` so that we know when build fails

File: build.gradle [hunk 10]
```diff
@@ -151,6 +151,8 @@ dependencyManagement {
 }
 
 dependencies {
+  compile group: 'com.microsoft.azure', name: 'azure-storage', version: '4.2.0'
```
Comment (nitinprabhu) [inline build.gradle]: Any reason we are not using 
`compile group: 'com.microsoft.azure', name: 'azure-storage', version: '7.0.0'
`

File: build.gradle [hunk 11]
```diff
@@ -171,12 +173,14 @@ dependencies {
...
+  smokeTestCompile sourceSets.test.runtimeClasspath
+  smokeTestCompile group: 'io.rest-assured', name: 'rest-assured', version: '3.0.7'
 }
 
-mainClassName = 'uk.gov.hmcts.reform.demo.Application'
+mainClassName = 'uk.gov.hmcts.reform.bulkscanning.Application'
 
 bootJar {
-  archiveName = 'spring-boot-template.jar'
+  archiveName = 'bulk-scanning.jar'
```
Comment (nitinprabhu) [inline build.gradle]: should we name the repo as `scanned-doc-processor` ? As actually it is not doing scanning ? May be good if we do it sooner before going into prod.

File: infrastructure/main.tf (hunk 2 above)
Comment (nitinprabhu) [reply infrastructure/main.tf]: it's not required.asp_name is the name of the shared service plan to be deployed to. Name should follow ${product}-${env}-asp format

File: infrastructure/main.tf [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+resource "azurerm_storage_account" "provider" {
+  name                     = "bulkscanning${var.env}"
+  resource_group_name      = "${azurerm_resource_group.rg.name}"
+  location                 = "${var.location}"
+  account_tier             = "Standard"
+  account_replication_type = "LRS"
+}
+
+resource "azurerm_storage_container" "incoming" {
+  name                  = "incoming"
```
Comment (nitinprabhu) [inline infrastructure/main.tf]: may be `scannedoccontainer` ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanning/info/BlobContainerInfoContributor.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    private final CloudBlobContainer container;
+
+    public BlobContainerInfoContributor(CloudBlobContainer container) {
+        this.container = container;
+    }
+
+    @Override
+    public void contribute(Builder builder) {
+        builder.withDetails(ImmutableMap.of(
+            "container_name", container.getName(),
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanning/info/BlobContainerInfoContributor.java]: we will loose build commit and date with this customization.Could you add it as an additional property to build info?

File: build.gradle (hunk 11 above)
Comment (KarolJastrzebski) [reply build.gradle]: :+1: Let's run this idea past the team and address in another PR.

File: infrastructure/main.tf (hunk 12 above)
Comment (KarolJastrzebski) [reply infrastructure/main.tf]: Can do that in another PR :+1:

**Pull request merged**